### ISSUE TITLE
Provide a way to include addition language definitions for highlight.js

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#2476]
The main highlight.js bundle only includes a portion of the languages that highlight.js supports. In the case the document uses a language not in this group, it should be possible for the user to specify additional definitions to be loaded from the CDN. (Even better, these languages can be auto-detected). Otherwise, the user is forced to download the highlight.js files and load them from disk.

If we decide to use an attribute, I propose something like `highlightjs-extra-langs`.

**Comment by mojavelinux** [comment]
This is a duplicate of #3036. We decided on `highlightjs-languages` since it could be extra languages or it could be all the languages. It depends on how highlight.js is loaded.

**Comment by bric3** [comment]
Hi,

Sorry to jump in on a closed issue, but I wonder how I can register a highlight language that is not part of the _regular_ highlight.js languages.

The issue #3086 only solves part fo the problem, you can activate [yaml](https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.13.1/languages/yaml.min.js), but not the non packageable languages like terraform.

To enable this terraform support. This likely support require a self download of the highlight.js library. But I don't see how it's possible to register the language as asciidoctor only includes highlight.js after the _docinfo-footer_.

```html
<script src="js/tocbot/tocbot.min.js"></script>
<script src="js/toc.js"></script>
<link rel="stylesheet" href="./js/highlight/styles/hl-dark-mode-aware.min.css">
<!-- end of docinfo-footer -->
<script src="./js/highlight/highlight.min.js"></script>
<script>hljs.initHighlighting()</script>
</body>
</html>
```

However to make this work the initialization functions needs to happen after highlight js is declared.

```html
<script src="./js/highlight/highlight.min.js"></script>
<script src="./js/highlight/terraform.js"></script>
<script type="text/javascript">
    hljs.registerLanguage('terraform', window.hljsDefineTerraform);
    hljs.initHighlighting()
</script>
```



----
asciidoctor 1.5.6
asciidoctor gralde plugin 2.4.0

**Comment by Mogztter** [comment]
@bric3 `docinfo-footer` should be appended after the syntax highlighter, see: 

https://github.com/asciidoctor/asciidoctor/blob/194d191f24dd6a069055c903cd7019772cfa988f/lib/asciidoctor/converter/html5.rb#L260
https://github.com/asciidoctor/asciidoctor/blob/194d191f24dd6a069055c903cd7019772cfa988f/lib/asciidoctor/converter/html5.rb#L296

> asciidoctor 1.6.2

You mean `asciidoctorj` right? I think you might need to upgrade to Asciidoctor 2.0.
It's not efficient but the following should work:

```html
<script src="./js/highlight/highlight.min.js"></script>
<script>hljs.initHighlighting()</script>
<!-- docinfo-footer -->
<script src="./js/highlight/terraform.js"></script>
<script type="text/javascript">
  hljs.registerLanguage('terraform', window.hljsDefineTerraform);
  document.querySelectorAll('pre code').forEach((block) => {
    hljs.highlightBlock(block);
  });
</script>
```

`initHighlighting` has a global state and will run only once, but you can call `highlightBlock` explicitly. I guess it should be possible to use a more precise query selector to reduce even more the overhead.

**Comment by Mogztter** [comment]
> The issue #3086 only solves part fo the problem

I think you are referring to #3036

**Comment by bric3** [comment]
@Mogztter Thanks for chiming in, it's very helpful, actually I checked the html backend in 1.5.8, and the doc info footer is appended before the highlight js block. That's why I couldn't make it work.


>>    asciidoctor 1.6.2
>
> You mean asciidoctorj right? I think you might need to upgrade to Asciidoctor 2.0.

So yes indeed I'll need to upgrade to make this work.

Fortunately I juste noticed that the gradle plugin that supports asciidoctorj 2 has been [released 4 days ago](https://github.com/asciidoctor/asciidoctor-gradle-plugin/releases/tag/release_3_0_0).

> I think you are referring to #3036

Yes, indeed, sorry for the typo.

**Comment by mojavelinux** [comment]
There's an even simpler way to do this since Asciidoctor 2. It's possible to register a replacement syntax highlighter that provides the docinfo that the built-in converter uses. You can even extend the built-in syntax highlighter so al you have to do is replace that method.

```
class MyHighlightJsAdapter < (Asciidoctor::SyntaxHighlighter.for 'highlight.js')
  register_for 'highlight.js'
  def docinfo location, doc, opts
    # load what you want
    '<script>hljs.initHighlighting()</script>'
  end
end
```

That's the whole reason the syntax highlighters have docinfo integration.

**Comment by bric3** [comment]
@mojavelinux Nice! I'm not familiar on how to do that with asciidoctorj / asciidoctor-gradle-plugin yet. Thanks for the tip.

**Comment by robertpanzer** [comment]
@bric3  maybe this helps with how implement this with AsciidoctorJ:  https://github.com/asciidoctor/asciidoctorj/blob/master/docs/integrator-guide.adoc

**Comment by bric3** [comment]
For reference I did it.

```
package com.github.bric3;

import org.asciidoctor.ast.Document;
import org.asciidoctor.extension.LocationType;
import org.asciidoctor.syntaxhighlighter.SyntaxHighlighterAdapter;

import java.util.Map;

/**
 * Provides a specific configuration of highlightjs
 *
 * https://github.com/asciidoctor/asciidoctorj/blob/master/docs/integrator-guide.adoc#implement-a-syntax-highlighter-adapter
 */
public class HighlightJsHighlighter implements SyntaxHighlighterAdapter {

    @Override
    public boolean hasDocInfo(LocationType location) {
        return location == LocationType.FOOTER;
    }

    @Override
    public String getDocinfo(LocationType location, Document document, Map<String, Object> options) {
        String highlightjsDir = ensureTrailingSlash(document.getAttribute("highlightjsdir").toString());
        Object theme = document.getAttribute("highlightjs-theme");
        return "<link rel=\"stylesheet\" href=\"" + highlightjsDir + "/styles/" + theme + ".min.css\">\n" +
                "<script src=\"" + highlightjsDir + "highlight.min.js\"></script>\n" +
                "<script src=\"" + highlightjsDir + "terraform.js\"></script>\n" +
                "<script>hljs.registerLanguage('terraform', window.hljsDefineTerraform);hljs.initHighlighting();</script>";
    }

    private String ensureTrailingSlash(String highlightjsDir) {
        return highlightjsDir.endsWith("/") ? highlightjsDir : highlightjsDir + "/";
    }
}
```

**Comment by mojavelinux** [comment]
Awesome!!

**Comment by bric3** [comment]
@mojavelinux Actually I do have an issue asciidoctor is set with a custom syntax highlighter extension, the generated html by asciidoc differs a bit so that highlight.js have issues to find the relevant language highlighter.

The generated HTML by asciidoctorj using **my custom highlightjs overiding extension** do not output the `language-java hljs` classes on the code element compared to standard highlightjs asciidoctor output.

```diff
+<pre class="highlightjs highlight"><code data-lang="java">
-<pre class="highlightjs highlight"><code data-lang="java" class="language-java hljs">
```

This lack of classes makes highligh.js guess what is the content and fails sometime, e.g. intead of java highlight.js select dart as the best match.

<details><summary>html snippets</summary>
* Generated HTML by asciidoctorj using **my custom highlightjs overiding extension**

    ```html
    <div class="listingblock">
    <div class="title">java</div>
    <div class="content">
    <pre class="highlightjs highlight"><code data-lang="java">public class HighlightJsHighlighter implements SyntaxHighlighterAdapter {
    
        @Override
        public boolean hasDocInfo(LocationType location) {
            return location == LocationType.FOOTER;
        }
    
        @Override
        public String getDocinfo(LocationType location, Document document, Map&lt;String, Object&gt; options) {
            String highlightjsDir = ensureTrailingSlash(document.getAttribute("highlightjsdir").toString());
            Object theme = document.getAttribute("highlightjs-theme");
            return "&lt;link rel=\"stylesheet\" href=\"" + highlightjsDir + "/styles/" + theme + ".min.css\"&gt;\n" +
                    "&lt;script src=\"" + highlightjsDir + "highlight.min.js\"&gt;&lt;/script&gt;\n" +
                    "&lt;script src=\"" + highlightjsDir + "terraform.js\"&gt;&lt;/script&gt;\n" +
                    "&lt;script&gt;hljs.registerLanguage('terraform', window.hljsDefineTerraform);hljs.initHighlighting();&lt;/script&gt;";
        }
    
        private String ensureTrailingSlash(String highlightjsDir) {
            return highlightjsDir.endsWith("/") ? highlightjsDir : highlightjsDir + "/";
        }
    }</code></pre>
    </div>
    </div>
    ```

    Notice the `pre code` classes : `<pre class="highlightjs highlight"><code data-lang="java" class="language-java hljs">`

* Generated HTML by asciidoctorj using the **standard highlightjs support**

    ```html
    <div class="listingblock">
    <div class="title">java</div>
    <div class="content">
    <pre class="highlightjs highlight"><code data-lang="java" class="language-java hljs">public class HighlightJsHighlighter implements SyntaxHighlighterAdapter {
    
        @Override
        public boolean hasDocInfo(LocationType location) {
            return location == LocationType.FOOTER;
        }
    
        @Override
        public String getDocinfo(LocationType location, Document document, Map&lt;String, Object&gt; options) {
            String highlightjsDir = ensureTrailingSlash(document.getAttribute("highlightjsdir").toString());
            Object theme = document.getAttribute("highlightjs-theme");
            return "&lt;link rel=\"stylesheet\" href=\"" + highlightjsDir + "/styles/" + theme + ".min.css\"&gt;\n" +
                    "&lt;script src=\"" + highlightjsDir + "highlight.min.js\"&gt;&lt;/script&gt;\n" +
                    "&lt;script src=\"" + highlightjsDir + "terraform.js\"&gt;&lt;/script&gt;\n" +
                    "&lt;script&gt;hljs.registerLanguage('terraform', window.hljsDefineTerraform);hljs.initHighlighting();&lt;/script&gt;";
        }
    
        private String ensureTrailingSlash(String highlightjsDir) {
            return highlightjsDir.endsWith("/") ? highlightjsDir : highlightjsDir + "/";
        }
    }</code></pre>
    </div>
    </div>
    ```
</details>
    

I'm not quite sure if I read this ruby code (v2.0.10) correctly

https://github.com/asciidoctor/asciidoctor/blob/9302213ea29e2efa75a7b3f9124ce4bbcd738174/lib/asciidoctor/converter/html5.rb#L652-L661

https://github.com/asciidoctor/asciidoctor/blob/9302213ea29e2efa75a7b3f9124ce4bbcd738174/lib/asciidoctor/converter/html5.rb#L666-L672

https://github.com/asciidoctor/asciidoctor/blob/9302213ea29e2efa75a7b3f9124ce4bbcd738174/lib/asciidoctor/syntax_highlighter/highlightjs.rb#L11-L13

But I have the feeling that the way this code is built the `format` is my custom syntax highlighter that uses the Java API, but since we cannot define the `format` method, then nothing fancy is output.

---- 
_Damn I wish github issues had asciidoc support!_

**Comment by bric3** [comment]
Ah ! I've got it, I should have the other part of syntax highlighting https://github.com/asciidoctor/asciidoctorj/blob/master/docs/integrator-guide.adoc#formatting-the-source-block-element, I just needed the formatter bit !

**Comment by mojavelinux** [comment]
>  I just needed the formatter bit !

The reason this is necessary is because you are writing one from scratch instead of extending the built-in one for highlight.js. I'm not sure if it's possible to extend the built-in one in AsciidoctorJ / Java. The code I provided above shows how to do it in Ruby.

**Comment by bric3** [comment]
@mojavelinux Yes that's what I grasped when reading the  javadoc of the Java API. Again thank you very much for the help !!!
